internal/bcdb/transaction_processor.go

```diff
@@ -43,78 +47,119 @@ type transactionProcessor struct {
-			BatchTimeout:       conf.batchTimeout,
+			MaxTxCountPerBatch: localConfig.BlockCreation.MaxTransactionCountPerBlock,
+			BatchTimeout:       localConfig.BlockCreation.BlockTimeout,
 			Logger:             conf.logger,
 		},
 	)
 
 	var err error
-	if p.blockCreator, err = blockcreator.New(
+
+	p.blockProcessor = blockprocessor.New(
+		&blockprocessor.Config{
+			BlockOneQueueBarrier: p.blockOneQueueBarrier,
+			BlockStore:           conf.blockStore,
+			ProvenanceStore:      conf.provenanceStore,
+			StateTrieStore:       conf.stateTrieStore,
+			DB:                   conf.db,
+			Logger:               conf.logger,
+		},
+	)
+
+	ledgerHeight, err := conf.db.Height()
+	if err != nil {
+		return nil, err
+	}
+	if ledgerHeight == 0 {
+		p.logger.Info("Bootstrapping the ledger and database")
+		tx, err := PrepareBootstrapConfigTx(conf.config)
+		if err != nil {
+			return nil, err
+		}
+		bootBlock, err := blockcreator.BootstrapBlock(tx)
+		if err != nil {
+			return nil, err
+		}
+		if err = p.blockProcessor.Bootstrap(bootBlock); err != nil {
+			return nil, err
+		}
+	}
+
+	p.blockCreator, err = blockcreator.New(
 		&blockcreator.Config{
 			TxBatchQueue: p.txBatchQueue,
 			Logger:       conf.logger,
 			BlockStore:   conf.blockStore,
 		},
-	); err != nil {
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	p.peerTransport = comm.NewHTTPTransport(&comm.Config{
+		LocalConf: localConfig,
+		Logger:    conf.logger,
+	})
+
+	clusterConfig, _, err := conf.db.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	conf.logger.Debugf("cluster config: %+v", clusterConfig)
+	if err = p.peerTransport.UpdateClusterConfig(clusterConfig); err != nil {
 		return nil, err
 	}
 
 	p.blockReplicator = replication.NewBlockReplicator(
 		&replication.Config{
+			LocalConf:            localConfig,
+			ClusterConfig:        clusterConfig,
+			Transport:            p.peerTransport,
 			BlockOneQueueBarrier: p.blockOneQueueBarrier,
 			Logger:               conf.logger,
 		},
 	)
-
+	if err = p.peerTransport.SetConsensusListener(p.blockReplicator); err != nil {
+		return nil, err
+	}
 	p.blockCreator.RegisterReplicator(p.blockReplicator)
 
-	p.blockProcessor = blockprocessor.New(
-		&blockprocessor.Config{
-			BlockOneQueueBarrier: p.blockOneQueueBarrier,
-			BlockStore:           conf.blockStore,
-			ProvenanceStore:      conf.provenanceStore,
-			StateTrieStore:       conf.stateTrieStore,
-			DB:                   conf.db,
-			Logger:               conf.logger,
-		},
-	)
-
-	_ = p.blockProcessor.RegisterBlockCommitListener(commitListenerName, p)
+	if err = p.blockProcessor.RegisterBlockCommitListener(commitListenerName, p); err != nil {
+		return nil, err
+	}
 
 	go p.txReorderer.Start()
 	p.txReorderer.WaitTillStart()
 
 	go p.blockCreator.Start()
 	p.blockCreator.WaitTillStart()
 
-	p.blockReplicator.Start()
+	p.peerTransport.Start() // Starts internal goroutine
+
+	p.blockReplicator.Start() // Starts internal goroutine
```

Comment:
Should we do a wait till start?

Reply:
This method waits until ready internally, we should probably refactor the others to do the same as well....